### PR TITLE
mm : Support multi-heap memory

### PR DESCRIPTION
**Pull request by jeongchanKim**
There are two cases to use multi-heap regions.
1. continuous physical RAM but want to split regions
2. separated physical RAMs but want to use for heap

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [d72cb284436792df16ad447f2eb023c221cc0f41] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d72cb284436792df16ad447f2eb023c221cc0f41] -

**Comment by seinfra** [comment]
Target : [45987c61a4196da314a76842d1a989cba4aec88e] - Code Rule Check OK.

File: os/mm/umm_heap/umm_malloc.c [hunk 1]
```diff
@@ -85,6 +85,33 @@
...
+ *   The address of the allocated memory (NULL on failure to allocate)
+ *
+ ************************************************************************/
+
+#if CONFIG_MM_NHEAPS > 1
+void *malloc_at(int heap_index, size_t size)
+{
+#ifdef CONFIG_DEBUG_MM_HEAPINFO
+	ARCH_GET_RET_ADDRESS
+	return mm_malloc(&g_mmheaps[heap_index], size, retaddr);
```
Comment (jeongarmy) [inline os/mm/umm_heap/umm_malloc.c]: I think you need to check heap_index to avoid assert.

File: os/arch/arm/src/common/up_allocateheap.c [hunk 2]
```diff
@@ -178,8 +182,13 @@ void up_allocate_heap(FAR void **heap_start, size_t *heap_size)
 
 	board_led_on(LED_HEAPALLOCATE);
 	*heap_start = (FAR void *)(g_idle_topstack & ~(0x7));
+#if CONFIG_MM_NHEAPS > 1
```
Comment (jeongarmy) [inline os/arch/arm/src/common/up_allocateheap.c]: needs <stdlib.h>

**Comment by seinfra** [comment]
Target : [ced729969ea550d1a44e52f26c9248f44b77407b] - Code Rule Check OK.

File: os/arch/Kconfig [hunk 3]
```diff
@@ -735,34 +736,60 @@ config RAM_VSTART
...
+	default "0,"
+	depends on MM_NHEAPS != 1
+	---help---
+		The size list of multi-heap memory.
+		Refer RAMx_START content. One different thing is this has
+		decimal values (START has hexa values).
+		This should have same number of list element with RAMx_START's.
+		Each size is separated by ',' as shown in example.
+		i.e. "100,200"
+
```
Comment (yashwanth686007) [inline os/arch/Kconfig]: Is there any condition check while initializing the heaps? i.e. whether the heap size is within the size range of region or not .

File: os/arch/arm/src/common/up_allocateheap.c [hunk 4]
```diff
@@ -205,15 +218,58 @@ void up_allocate_kheap(FAR void **heap_start, size_t *heap_size)
 }
 #endif
 
+/****************************************************************************
+ * Name: up_addmultiheap
+ ****************************************************************************/
+#if CONFIG_MM_NHEAPS > 1
```
Comment (thapav) [inline os/arch/arm/src/common/up_allocateheap.c]: #ifdef usage is recommended.
Please check all the #if 's

Comment (jeongchanKim) [reply os/arch/arm/src/common/up_allocateheap.c]: We should check not only that CONFIG_MM_NHEAPS is defined, but also the value of this config.
So I think `#if CONFIG_MM_NHEAPS > 1` is correct.

Comment (thapav) [reply os/arch/arm/src/common/up_allocateheap.c]: ahh yes

File: os/mm/umm_heap/umm_malloc.c [hunk 5]
```diff
@@ -85,6 +86,37 @@
...
+ * Return Value:
+ *   The address of the allocated memory (NULL on failure to allocate)
+ *
+ ************************************************************************/
+
+#if CONFIG_MM_NHEAPS > 1
+void *malloc_at(int heap_index, size_t size)
+{
+	if (heap_index >= CONFIG_MM_NHEAPS) {
+		mdbg("Wrong heap index, so Allocation Failed.\n");
```
Comment (sunghan-chang) [inline os/mm/umm_heap/umm_malloc.c]: For debugging information, let's print `heap_index` and `CONFIG_MM_NHEAPS`.

File: os/mm/umm_heap/umm_initialize.c [hunk 6]
```diff
@@ -74,8 +74,12 @@
 #if !defined(CONFIG_ARCH_ADDRENV) || !defined(CONFIG_BUILD_KERNEL)
 /* Otherwise, the user heap data structures are in common .bss */
 
+#if CONFIG_MM_NHEAPS > 1
+struct mm_heap_s g_mmheaps[CONFIG_MM_NHEAPS] = {0};
```
Comment (sunghan-chang) [inline os/mm/umm_heap/umm_initialize.c]: Original code does not initialize with zero. Is this necessary?

File: os/arch/arm/src/common/up_initialize.c [hunk 7]
```diff
@@ -166,6 +166,10 @@ void up_initialize(void)
 
 	up_color_intstack();
 
+	/* Add extra heap memory fragments for multi-heap memory manager */
+
+	up_addmultiheap();
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_initialize.c]: Is this right to add multi heap setting before adding region?

File: os/arch/Kconfig [hunk 8]
```diff
@@ -735,34 +736,60 @@ config RAM_VSTART
...
+	depends on MM_NHEAPS != 1
+	---help---
+		The address list of the multiple heap memory.
+		When you want to use multi-heap in a RAM, set MM_NHEAPS first at memory
+		management, and then set hexa values at RAMx_START and RAMx_SIZE
+		to support multi-heap memory.
+		Each address is separated by ',' as shown in example.
+		i.e. "0x02000000,0x4000000"
+
+config RAMx_SIZE
```
Comment (sunghan-chang) [inline os/arch/Kconfig]: We had better change a name to clearly say the meaning of configs like PHI_RAMx_SIZE.

File: os/arch/arm/src/common/up_allocateheap.c [hunk 9]
```diff
@@ -178,8 +186,13 @@ void up_allocate_heap(FAR void **heap_start, size_t *heap_size)
 
 	board_led_on(LED_HEAPALLOCATE);
 	*heap_start = (FAR void *)(g_idle_topstack & ~(0x7));
+#if CONFIG_MM_NHEAPS > 1
+	*heap_size =  (size_t)strtol(CONFIG_RAMx_SIZE, NULL, 0);
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_allocateheap.c]: wrong size.

File: os/arch/arm/src/common/up_allocateheap.c [hunk 10]
```diff
@@ -205,15 +218,58 @@ void up_allocate_kheap(FAR void **heap_start, size_t *heap_size)
 }
 #endif
 
+/****************************************************************************
+ * Name: up_addmultiheap
+ ****************************************************************************/
+#if CONFIG_MM_NHEAPS > 1
+void up_addmultiheap(void)
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_allocateheap.c]: up_add_extra_heapx(void)

File: os/include/tinyara/mm/mm.h [hunk 11]
```diff
@@ -384,9 +384,12 @@ extern "C" {
  *   structure is associated with the address environment and there is
  *   no global user heap structure.
  */
-
+#if CONFIG_MM_NHEAPS > 1
+EXTERN struct mm_heap_s g_mmheaps[CONFIG_MM_NHEAPS];
+#else
```
Comment (sunghan-chang) [inline os/include/tinyara/mm/mm.h]: Let's remove `#else`

File: os/mm/umm_heap/umm_malloc.c (hunk 1 above)
Comment (shadaksharikp) [reply os/mm/umm_heap/umm_malloc.c]: i feel Similarly Size argument limits validation also needed.
What is your opinion ?

Comment (jeongchanKim) [reply os/mm/umm_heap/umm_malloc.c]: size will be checked in mm_malloc.

**Comment by seinfra** [comment]
Target : [fb22f4ee52cf78183f37dd4789c423d40231d89b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fb22f4ee52cf78183f37dd4789c423d40231d89b] - Code Rule Check (C++) OK.

**Comment by sunghan-chang** [comment]
@jeongchanKim As we discussed on offline, let's re-implement this logic fully, not following Nuttx structure.

**Comment by seinfra** [comment]
Target : [f6b8bf284d912ba8ccc3182d4e62d7fc438f6e5c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f6b8bf284d912ba8ccc3182d4e62d7fc438f6e5c] - Code Rule Check OK.

File: os/tools/mk_heap_regioninfo.sh [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+#
+# File : mk_heap_regioninfo.sh
+
+WD=`test -d ${0%/*} && cd ${0%/*}; pwd`
```
Comment (sunghan-chang) [inline os/tools/mk_heap_regioninfo.sh]: WD -> OSTOOLSDIR

File: os/tools/mk_heap_regioninfo.sh [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+TOPDIR="${WD}/.."
```
Comment (sunghan-chang) [inline os/tools/mk_heap_regioninfo.sh]: TOPDIR -> OSDIR

**Comment by seinfra** [comment]
Target : [e78b8696e49b99a286b7be377437014f80c8c4f2] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e78b8696e49b99a286b7be377437014f80c8c4f2] - Code Rule Check (C++) OK.

File: os/arch/Kconfig [hunk 14]
```diff
@@ -735,37 +736,50 @@ config RAM_VSTART
...
 	---help---
-		The start address list of the additional heap regions.
-		When you want to use separated heap region in a physical RAM or
-		two different physical RAMs, set MM_REGIONS first at memory
-		management, and then set hexa values at RAMx_START and RAMx_SIZE
+		The address list of the RAM regions.
+		If you want to use more than two RAMs physically, or
+		if you want to divide RAM into multiple heaps,
+		set MM_REGIONS first at memory management, and then
+		set hexa values at REGIONx_START and REGIONx_SIZE
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/arch/Kconfig]: done.

File: os/arch/arm/src/common/up_allocateheap.c [hunk 15]
```diff
@@ -67,7 +67,16 @@
 #include <arch/board/board.h>
 
 #if CONFIG_MM_REGIONS > 1
+#include <tinyara/mm/mm.h>
 #include <tinyara/kmalloc.h>
+#include <stdlib.h>
+#endif
+
+#if CONFIG_MM_NHEAPS > 1
+#include <tinyara/mm/heap_regioninfo.h>
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/arch/arm/src/common/up_allocateheap.c]: I added __XXX definition. But externs are needed.

File: os/include/tinyara/mm/mm.h [hunk 16]
```diff
@@ -408,7 +407,7 @@ EXTERN struct mm_heap_s g_kmmheap;
 #else
 /* Otherwise, the user heap data structures are in common .bss */
 
-#define USR_HEAP &g_mmheap
+#define USR_HEAP &g_mmheap[0]
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/include/tinyara/mm/mm.h]: no. this is correct.

File: os/include/tinyara/mm/mm.h [hunk 17]
```diff
@@ -660,6 +659,10 @@ void mm_is_sem_available(void);
 /* Functions to get heap information */
 struct mm_heap_s *mm_get_heap_info(void);
 
+#if CONFIG_MM_NHEAPS > 1
+void *malloc_at(int heap_index, size_t size);
+#endif
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/include/tinyara/mm/mm.h]: done.

File: os/tools/mk_heap_regioninfo.sh [hunk 18]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/tools/mk_heap_regioninfo.sh]: done.

**Comment by seinfra** [comment]
Target : [e30e4bd92f92f40d2064afc12d678db5bd51fa4c] -

**Comment by seinfra** [comment]
Target : [e30e4bd92f92f40d2064afc12d678db5bd51fa4c] -

**Comment by seinfra** [comment]
Target : [026ea7752200505b1e136592357e4c2098f75afa] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [026ea7752200505b1e136592357e4c2098f75afa] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [31025a1084bfb2582d4ed670f1a8068052386ea6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [31025a1084bfb2582d4ed670f1a8068052386ea6] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4aadd0dbf1d89b59a8606c28035b830ea40dff72] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [4aadd0dbf1d89b59a8606c28035b830ea40dff72] - Code Rule Check OK.

File: docs/HowToUseMultiHeap.md [hunk 19]
```diff
@@ -1,30 +1,54 @@
...
-1. continuous physical RAM but want to split regions  
+There are two cases to use multi-heap regions.
+1. continuous physical RAM but want to split regions
 2. separated physical RAMs but want to use for heap  
 
-To use multi-heap regions, find below two steps.
+To use multi-heap, find below two steps.
 
-## Set number of Heap regions
+## Set number of Regions
```
Comment (sunghan-chang) [inline docs/HowToUseMultiHeap.md]: Need a description of difference between `region` and `heap`

File: os/mm/Kconfig [hunk 20]
```diff
@@ -64,9 +64,17 @@ config MM_REGIONS
...
+		memory. Or if you want to use RAM to multiple region(heap), MM_REGIONS
+		should be set to the number of heaps.
+
+config MM_NHEAPS
+	int "Number of heaps"
+	default 1
+	---help---
+		A number of heap memory. If MM_NHEAPS sets to 1, RAM memory can be
+		used for one heap memory. If MM_NHEAPS sets more than equal to 2, it can be
+		split for multi-heap.
```
Comment (sunghan-chang) [inline os/mm/Kconfig]: Could you leave a link of README here?

File: os/tools/mk_heap_regioninfo.sh [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+REGIONx_START=${CONFIG_RAM_REGIONx_START}
+REGIONx_SIZE=${CONFIG_RAM_REGIONx_SIZE}
+REGIONx_INDEX=${CONFIG_RAM_REGIONx_INDEX}
+
+REGION_NUM=$(printf ${REGIONx_START} | grep -o ',' | wc -l)
+REGION_NUM=$((REGION_NUM+1))
+
+printf '%s\n' "#define MM_REGION_NUM ${REGION_NUM}"> ${OSDIR}/include/tinyara/mm/heap_regioninfo.h
+
+printf '%s\n' "#ifndef __TINYARA_MM_HEAP_REGIONINFO_H"> ${INTERNAL_H_PATH}
```
Comment (sunghan-chang) [inline os/tools/mk_heap_regioninfo.sh]: 1. definition and file name are mis-matched.
2. let's rename those files name

File: os/tools/mkconfig.c [hunk 22]
```diff
@@ -206,14 +206,17 @@ int main(int argc, char **argv, char **envp)
 	printf("#ifndef CONFIG_MM_REGIONS\n");
 	printf("# define CONFIG_MM_REGIONS 1\n");
 	printf("#endif\n\n");
+	printf("#if CONFIG_MM_REGIONS > 1\n");
```
Comment (sunghan-chang) [inline os/tools/mkconfig.c]: Let's remove this and add a converting string to int in shell script.

**Comment by sunghan-chang** [comment]
@manoharahk @yashwanth686007 @thapav Could you review this? Do you understand what it is from README?

File: docs/HowToUseMultiHeap.md (hunk 19 above)
Comment (jeongchanKim) [reply docs/HowToUseMultiHeap.md]: done

**Comment by seinfra** [comment]
Target : [36fbd7c2487ca500f25a030aed7bb8ed741e3343] -

**Comment by seinfra** [comment]
Target : [581826a34e27ac41c4ac10d8f49622edf0127849] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [581826a34e27ac41c4ac10d8f49622edf0127849] - Code Rule Check OK.

File: docs/HowToUseMultiHeap.md [hunk 23]
```diff
@@ -1,30 +1,57 @@
...
-Hardware Configuration -> Chip Selection -> List of size for additional heap region -> set values
+Memory Management -> Number of heaps -> change a number over 2
+```
+
+<U>For a detailed description of **REGION** and **HEAP**, please see the link below.</U>  
+[Terms for REGION and HEAP](#example-of-multi-heap-usage)  
+
+## Set Region start address, size and index
+
+Set start addresses, **CONFIG_RAM_REGIONx_START** with hexa values,  set sizes, **CONFIG_RAM_REGIONx_SIZE** with decimal values of new heap, and set indexes, **CONFIG_RAM_REGIONx_INDEX** with decimal values of new heap.
```
Comment (yashwanth686007) [inline docs/HowToUseMultiHeap.md]: I think its better to mention that decimal value should be in bytes.

File: docs/HowToConfigureMemory.md [hunk 24]
```diff
@@ -17,10 +17,10 @@ Find configs with following step:
 ```
 Hardware Configuration -> Chip selection -> Boot Memory Configuration
 ```
-Set CONFIG_RAM_START as hexa value and set CONFIG_RAM_SIZE as decimal value.
+Set CONFIG_RAM_REGIONx_START as hexa value and set CONFIG_RAM_SIZE as decimal value.
```
Comment (yashwanth686007) [inline docs/HowToConfigureMemory.md]: Change CONFIG_RAM_SIZE -> CONFIG_RAM_REGIONx_SIZE

File: os/mm/umm_heap/umm_malloc.c [hunk 25]
```diff
@@ -85,6 +86,37 @@
...
+ *
+ * Return Value:
+ *   The address of the allocated memory (NULL on failure to allocate)
+ *
+ ************************************************************************/
+
+#if CONFIG_MM_NHEAPS > 1
+void *malloc_at(int heap_index, size_t size)
+{
+	if (heap_index >= CONFIG_MM_REGIONS) {
```
Comment (yashwanth686007) [inline os/mm/umm_heap/umm_malloc.c]: Also check for negative value of heap index.

File: os/include/tinyara/mm/mm.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (yashwanth686007) [reply os/include/tinyara/mm/mm.h]: @jeongchanKim , I think both represent the value?

File: os/mm/Kconfig (hunk 20 above)
Comment (jeongchanKim) [reply os/mm/Kconfig]: done.

**Comment by seinfra** [comment]
Target : [6d41acea8affdad48654fe87982f8f18859f1283] - Code Rule Check Result:
os/arch/arm/src/common/up_allocateheap.c:231: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'

**Comment by seinfra** [comment]
Target : [6d41acea8affdad48654fe87982f8f18859f1283] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f855abeeee8739e605e3f809fc0fa91e4d1c1175] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f855abeeee8739e605e3f809fc0fa91e4d1c1175] - Code Rule Check OK.

File: docs/HowToConfigureMemory.md [hunk 26]
```diff
@@ -17,10 +17,10 @@ Find configs with following step:
 ```
 Hardware Configuration -> Chip selection -> Boot Memory Configuration
 ```
-Set CONFIG_RAM_START as hexa value and set CONFIG_RAM_SIZE as decimal value.
+Set CONFIG_RAM_REGIONx_START as hexa value and set CONFIG_RAM_REGIONx_SIZE as decimal value.
```
Comment (sunghan-chang) [inline docs/HowToConfigureMemory.md]: Please write "in bytes" here as well.

File: os/Makefile.unix [hunk 27]
```diff
@@ -295,6 +295,7 @@ endif
 # tools/mkversion tool is built and used to create include/tinyara/version.h
 
 force_build:
+	$(Q) $(TOPDIR)/tools/mk_heap_regioninfo.sh
```
Comment (sunghan-chang) [inline os/Makefile.unix]: Please make a new tab to execute this like version and config.

**Comment by seinfra** [comment]
Target : [b3fa02f3738473ee653adedfb7c37e08f45a4c78] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [b3fa02f3738473ee653adedfb7c37e08f45a4c78] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [553e91641408f2fbe049bac7e15c78466b03bcda] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [49f345d24cb59ac0c14e9acc45c309838f5130db] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1baac670986fad7524041625a612522029041a2f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1baac670986fad7524041625a612522029041a2f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [f6ddc15001b260d3deecfddfb8514cbdd55e1a93] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f6ddc15001b260d3deecfddfb8514cbdd55e1a93] - Code Rule Check OK.

File: os/Makefile.unix (hunk 27 above)
Comment (jeongchanKim) [reply os/Makefile.unix]: done.

File: os/include/tinyara/mm/mm.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/include/tinyara/mm/mm.h]: with &g_mmheap, it will be error when build time.

**Comment by seinfra** [comment]
Target : [3c325825d03cd34b4c0fd4011a82773e9a1385e3] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3c325825d03cd34b4c0fd4011a82773e9a1385e3] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [93d46e2b0bb784e322d0e65c7e6bc13575a60fff] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [93d46e2b0bb784e322d0e65c7e6bc13575a60fff] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1d9b1b12e68590ca5814f67b4c68c2f6d3015020] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1d9b1b12e68590ca5814f67b4c68c2f6d3015020] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fa07c6c6c562d5d998aaaa495ef746606836f2d7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fa07c6c6c562d5d998aaaa495ef746606836f2d7] - Code Rule Check OK.

File: os/arch/arm/src/common/up_allocateheap.c [hunk 28]
```diff
@@ -66,9 +66,15 @@
...
+#include <tinyara/mm/mm.h>
 #include <tinyara/kmalloc.h>
-#endif
+#include <stdlib.h>
+#include <stdbool.h>
+extern void *regionx_start[CONFIG_MM_REGIONS];
+extern size_t regionx_size[CONFIG_MM_REGIONS];
+extern int regionx_heap_idx[CONFIG_MM_REGIONS];
+
+bool init_region[CONFIG_MM_NHEAPS];
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_allocateheap.c]: regionx_is_init

File: os/arch/arm/src/common/up_allocateheap.c [hunk 29]
```diff
@@ -66,9 +66,15 @@
 
 #include <arch/board/board.h>
 
-#if CONFIG_MM_REGIONS > 1
+#include <tinyara/mm/mm.h>
 #include <tinyara/kmalloc.h>
-#endif
+#include <stdlib.h>
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_allocateheap.c]: let's locate them after stdint.h

File: os/arch/arm/src/common/up_allocateheap.c [hunk 30]
```diff
@@ -161,9 +167,9 @@ void up_allocate_heap(FAR void **heap_start, size_t *heap_size)
 
 	uintptr_t ubss_start = (uintptr_t)USERSPACE->us_bssstart;
 	uintptr_t ubase = (uintptr_t)USERSPACE->us_bssend;
-	size_t usize = CONFIG_RAM_END - ubase;
+	size_t usize = (size_t)regionx_start[0] + regionx_size[0] - ubase;
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_allocateheap.c]: How about below
```
#define REGION_END(start,size)   (start) + (size)
```

**Comment by seinfra** [comment]
Target : [e808ec3221695127239a19b430ffe6f0fd39dc03] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e808ec3221695127239a19b430ffe6f0fd39dc03] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3dfe629dc97e88b7269c575cd1ea968b9f8c59a5] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [3dfe629dc97e88b7269c575cd1ea968b9f8c59a5] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [683d1b36fda07de08ddb1cf4c381bb8566f6a7ec] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [683d1b36fda07de08ddb1cf4c381bb8566f6a7ec] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1db5f765c7dc417a9819fba69c48ac946e0aec2a] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1db5f765c7dc417a9819fba69c48ac946e0aec2a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [62d78865f7dba47b132a61928d1777f310926e75] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [62d78865f7dba47b132a61928d1777f310926e75] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [02c7b887842e4b1b159947598b8f46bb6e974f1e] -

**Comment by seinfra** [comment]
Target : [00bc42bfd7e580df5b583f7f36a671c602ca549d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [00bc42bfd7e580df5b583f7f36a671c602ca549d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [9d3343025e216d8d008fd7abab6761963125b692] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4541e17d777fd44fbdac5c9b42c87c22e81b3712] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4541e17d777fd44fbdac5c9b42c87c22e81b3712] - Code Rule Check (C++) OK.

File: os/mm/Makefile [hunk 31]
```diff
@@ -92,7 +92,7 @@ KBIN = libkmm$(LIBEXT)
 BIN ?= libmm$(LIBEXT)
 
 all: $(BIN)
-.PHONY: clean distclean
+.PHONY: clean distclean include/tinyara/mm/heap.h
```
Comment (sunghan-chang) [inline os/mm/Makefile]: `heap_regioninfo` but I think this is not necessary. Could you check?

**Comment by sunghan-chang** [comment]
@jeongchanKim Could you check and let me know?
1. If we use `malloc`, not `malloc_at`, does it use heap 0 index?
2. When heap 0 index is full, how does it work? use next index?

**Comment by seinfra** [comment]
Target : [f1af6bccf1ce792b24ba67e04a58eb60bddd48a8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f1af6bccf1ce792b24ba67e04a58eb60bddd48a8] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [0a7460c64128ae1f8a73da7f2771c175fc2580fc] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [0a7460c64128ae1f8a73da7f2771c175fc2580fc] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [baa0075514aad395e7d03e5f5768b725fdaaf1e8] -

**Comment by seinfra** [comment]
Target : [baa0075514aad395e7d03e5f5768b725fdaaf1e8] -

**Comment by seinfra** [comment]
Target : [9b6411360c75d7031ad5a45a22ab8965a1d1747d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [9b6411360c75d7031ad5a45a22ab8965a1d1747d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [7c62b46bbca0faff0f2d20ce1707e9faaf8f5f79] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [157c67d077149312d80d18a8b2519866a1d3180f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [157c67d077149312d80d18a8b2519866a1d3180f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [25a6dc4f42b30733d501627ae6cc86f19b2f928b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [25a6dc4f42b30733d501627ae6cc86f19b2f928b] - Code Rule Check OK.

File: os/tools/mk_heap_regioninfo.sh [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+REGION_INFO_C_FILE_PATH=${OSDIR}/mm/mm_heap/mm_heap_regioninfo.c
+
+REGIONx_START=${CONFIG_RAM_REGIONx_START}
+REGIONx_SIZE=${CONFIG_RAM_REGIONx_SIZE}
+REGIONx_HEAP_IDX=${CONFIG_RAM_REGIONx_HEAP_INDEX}
+
+REGION_NUM=${CONFIG_MM_REGIONS}
+HEAP_NUM=${CONFIG_MM_NHEAPS}
+
+printf '%s\n%s\n\n%s' "#include <tinyara/config.h>" "#include <sys/types.h>" "void *regionx_start[CONFIG_MM_REGIONS] = {"> ${REGION_INFO_C_FILE_PATH}
```
Comment (sunghan-chang) [inline os/tools/mk_heap_regioninfo.sh]: At start of file, please add a caution sentence.
```
printf '%s\n' "/* Automatically generated file; DO NOT EDIT. */">${REGION_INFO_C_FILE_PATH}
```

File: os/mm/umm_heap/umm_sem.c [hunk 33]
```diff
@@ -97,7 +97,7 @@
 
 int umm_trysemaphore(void)
 {
-	return mm_trysemaphore(USR_HEAP);
+	return mm_trysemaphore(BASE_HEAP);
```
Comment (sunghan-chang) [inline os/mm/umm_heap/umm_sem.c]: need to discuss.
When we split heap regions, should we block to use heap for all?

**Comment by seinfra** [comment]
Target : [2092f7e1c79d6c9a8f27443f951add3fa828e090] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [b8fc909bf7189e5fcdca6cc4e0b227b953e4c546] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [b8fc909bf7189e5fcdca6cc4e0b227b953e4c546] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [25f8f78404ca8dc2c2922cdd0a88cc598b1c9bd8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [25f8f78404ca8dc2c2922cdd0a88cc598b1c9bd8] - Code Rule Check OK.

**Pull request merged**